sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
   return ret_err;
 }
 
+// USM
+pi_result cuda_piextUSMHostAlloc(void **result_ptr, pi_context context,
+                                 pi_usm_mem_properties *properties, size_t size,
+                                 pi_uint32 alignment) {
+  PI_CHECK_ERROR(cuMemAllocHost(result_ptr, size));
+  assert(*result_ptr % alignment == 0);
+  return PI_SUCCESS;
+}
+
+pi_result cuda_piextUSMDeviceAlloc(void **result_ptr, pi_context context,
+                                   pi_device device,
+                                   pi_usm_mem_properties *properties,
+                                   size_t size, pi_uint32 alignment) {
+  assert(device != nullptr);
+  ScopedContext active(context);
+  PI_CHECK_ERROR(cuMemAlloc((CUdeviceptr *)result_ptr, size));
+  assert(*result_ptr % alignment == 0);
+  return PI_SUCCESS;
+}
+
+pi_result cuda_piextUSMSharedAlloc(void **result_ptr, pi_context context,
+                                   pi_device device,
+                                   pi_usm_mem_properties *properties,
+                                   size_t size, pi_uint32 alignment) {
+  cl::sycl::detail::pi::die("cuda_piextUSMSharedAlloc not implemented");
+  return PI_ERROR_UNKNOWN;
+}
+
+pi_result cuda_piextUSMFree(pi_context context, void *ptr) {
+  pi_result retErr = PI_SUCCESS;
+  unsigned int type;
+  retErr = PI_CHECK_ERROR(cuPointerGetAttribute(
+      &type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE, (CUdeviceptr)ptr));
+  assert(type == CU_MEMORYTYPE_DEVICE || type == CU_MEMORYTYPE_HOST);
```

Comment:
Seems to me an assert is not enough in this case, this should be a proper PI_INVALID_VALUE error returned at runtime in case is none of the two.

Reply:
According to the documentation for [for `cuPointerGetAttribute`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__UNIFIED.html#group__CUDA__UNIFIED_1g0c28ed0aff848042bc0533110e45820c) these should be the only possible values:

> #### `CU_POINTER_ATTRIBUTE_MEMORY_TYPE`:
> 
> Returns in `*data` the physical memory type of the memory that `ptr` addresses as a `CUmemorytype` enumerated value. The type of data must be `unsigned int`.
> 
> If `ptr` addresses device memory then `*data` is set to `CU_MEMORYTYPE_DEVICE`. The particular `CUdevice` on which the memory resides is the `CUdevice` of the `CUcontext` returned by the `CU_POINTER_ATTRIBUTE_CONTEXT` attribute of `ptr`.
> 
> If `ptr` addresses host memory then `*data` is set to `CU_MEMORYTYPE_HOST`.
> 
> If `ptr` was not allocated by, mapped by, or registered with a `CUcontext` which uses unified virtual addressing then `CUDA_ERROR_INVALID_VALUE` is returned.
> 
> If the current `CUcontext` does not support unified virtual addressing then `CUDA_ERROR_INVALID_CONTEXT` is returned.

On the other hand, the documentation for [`enum CUmemorytype`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1g8a114cc994ad2e865c44ef3838eaec72) does list the additional values `CU_MEMORYTYPE_ARRAY` and `CU_MEMORYTYPE_UNIFIED`, so I can look into it a bit more.

Reply:
I think the test for shared will involve CU_POINTER_ATTRIBUTE_IS_MANAGED

Reply:
here's what I got from an explicit test:
```
Found 1 CUDA devices
Using device GeForce GTX 960M
cuMemAllocHost() returns a pointer of type CU_MEMORYTYPE_HOST
cuMemHostAlloc() returns a pointer of type CU_MEMORYTYPE_HOST
cuMemAlloc() returns a pointer of type CU_MEMORYTYPE_DEVICE
cuMemAllocManaged() returns a pointer of type CU_MEMORYTYPE_DEVICE
```

Reply:
which helps, since the memory returned by `cuMemAllocManaged()` should be freed by `cuMemFree()` as well.

Reply:
there is also [`cuMemHostRegister`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1gf0a9fe11544326dabd743b7aa6b54223), which from what I can _infer_ from the docs, returns `CU_MEMORYTYPE_HOST`, but I would expect `CU_MEMORYTYPE_UNIFIED`. I find that a bit confusing.
I guess an assert is OK for this, if the other values of the struct are not returned.

Reply:
I think cuMemHostRegister only pins and maps existing c/c++ allocations so they can be visible like "host" allocations.

Reply:
OK, I tested it and
```
cuMemHostRegister() returns a pointer of type CU_MEMORYTYPE_HOST
```

However, that pointer **cannot** be freed by `cuMemFreeHost`:
```
cudaCheck(cuMemFreeHost(buffer));
CUDA_ERROR_INVALID_VALUE: invalid argument
```

Reply:
>  I guess an assert is OK for this, if the other values of the struct are not returned.

OK, I'll switch back to an assert.

Reply:
Correct - this is memory that _was not allocated by cuda_.

The programmer should `cuMemHostUnregister` then deallocate with either `std::free` or `delete`.